// @cycle/history/lib/cjs/drivers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var history_1 = require("history");
var createHistory_1 = require("./createHistory$");
function isHistory(x) {
    return !!x.createHref;
}
function makeHistoryDriver(options) {
    var history;
    if (options && isHistory(options)) {
        history = options;
    }
    else {
        history = history_1.createBrowserHistory(options);
    }
    return function historyDriver(sink$) {
        return createHistory_1.createHistory$(history, sink$);
    };
}
exports.makeHistoryDriver = makeHistoryDriver;
function makeServerHistoryDriver(options) {
    var history = history_1.createMemoryHistory(options);
    return function serverHistoryDriver(sink$) {
        return createHistory_1.createHistory$(history, sink$);
    };
}
exports.makeServerHistoryDriver = makeServerHistoryDriver;
function makeHashHistoryDriver(options) {
    var history = history_1.createHashHistory(options);
    return function hashHistoryDriver(sink$) {
        return createHistory_1.createHistory$(history, sink$);
    };
}
exports.makeHashHistoryDriver = makeHashHistoryDriver;
